File: cpp/tests/test_mirror.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <algorithm>
#include <chrono>
#include <filesystem>
#include <string>
#include <thread>

namespace fs = std::filesystem;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

static fs::path make_temp_mirror_dir() {
    auto tmp = fs::temp_directory_path() /
               ("vost_mirror_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_mirror_store(const fs::path& path,
                                        const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

static bool contains(const std::vector<std::string>& v, const std::string& s) {
    return std::find(v.begin(), v.end(), s) != v.end();
}

static bool any_ref_contains(const std::vector<vost::RefChange>& changes,
                             const std::string& substr) {
    return std::any_of(changes.begin(), changes.end(),
                       [&](const vost::RefChange& r) {
                           return r.ref_name.find(substr) != std::string::npos;
                       });
}

// ---------------------------------------------------------------------------
// backup
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: backup to local bare repo", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();

    auto diff = store.backup(remote_url);

    CHECK_FALSE(diff.in_sync());
    CHECK_FALSE(diff.add.empty());

    // Verify remote has the refs
    auto remote = vost::GitStore::open(remote_path);
    auto branches = remote.branches().keys();
    CHECK(contains(branches, "main"));
    CHECK(remote.branches()["main"].read_text("a.txt") == "hello");

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

// ---------------------------------------------------------------------------
// restore
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: restore from local bare repo", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    // Create a new empty store and restore into it
    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions opts;
    opts.create = true;
    auto store2 = vost::GitStore::open(restore_path, opts);

    auto diff = store2.restore(remote_url);
    CHECK_FALSE(diff.in_sync());
    CHECK_FALSE(diff.add.empty());

    auto branches = store2.branches().keys();
    CHECK(contains(branches, "main"));
    CHECK(store2.branches()["main"].read_text("a.txt") == "hello");

    fs::remove_all(path);
    fs::remove_all(remote_path);
    fs::remove_all(restore_path);
}

// ---------------------------------------------------------------------------
// dry-run
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: dry-run backup makes no changes", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();

    // First do a real backup
    store.backup(remote_url);

    // Write more data
    f = store.branches()["main"];
    f = f.write_text("b.txt", "world");

    // Dry-run should report changes but not push
    vost::BackupOptions bo;
    bo.dry_run = true;
    auto diff = store.backup(remote_url, bo);
    CHECK_FALSE(diff.in_sync());

    // Remote should still only have the old data
    auto remote = vost::GitStore::open(remote_path);
    CHECK_FALSE(remote.branches()["main"].exists("b.txt"));

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

TEST_CASE("Mirror: dry-run restore makes no changes", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    // Create empty store
    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions opts;
    opts.create = true;
    auto store2 = vost::GitStore::open(restore_path, opts);

    vost::RestoreOptions ro;
    ro.dry_run = true;
    auto diff = store2.restore(remote_url, ro);
    CHECK_FALSE(diff.in_sync());

    // Store2 should still be empty
    CHECK(store2.branches().keys().empty());

    fs::remove_all(path);
    fs::remove_all(remote_path);
    fs::remove_all(restore_path);
}

// ---------------------------------------------------------------------------
// stale ref deletion (backup) / additive restore
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: backup deletes stale remote refs", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    // Create a second branch
    store.branches().set("extra", f);

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    // Verify remote has both branches
    {
        auto remote = vost::GitStore::open(remote_path);
        CHECK(contains(remote.branches().keys(), "extra"));
    }

    // Delete the extra branch locally
    store.branches().del("extra");

    // Backup again — should delete the remote extra branch
    auto diff = store.backup(remote_url);
    CHECK(any_ref_contains(diff.del, "extra"));

    // Verify remote no longer has the extra branch
    auto remote = vost::GitStore::open(remote_path);
    CHECK_FALSE(contains(remote.branches().keys(), "extra"));

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

TEST_CASE("Mirror: restore is additive", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    // Create a local-only branch
    store.branches().set("local-only", f);
    CHECK(contains(store.branches().keys(), "local-only"));

    // Restore from remote — local-only branch should survive (additive)
    auto diff = store.restore(remote_url);
    CHECK(diff.del.empty());
    CHECK(contains(store.branches().keys(), "local-only"));

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

// ---------------------------------------------------------------------------
// round-trip
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: round-trip backup then restore", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "aaa");
    f = f.write_text("b.txt", "bbb");

    store.branches().set("feature", f);
    auto feat = store.branches()["feature"];
    feat = feat.write_text("c.txt", "ccc");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    // Create new store and restore
    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions opts;
    opts.create = true;
    auto store2 = vost::GitStore::open(restore_path, opts);
    store2.restore(remote_url);

    CHECK(store2.branches()["main"].read_text("a.txt") == "aaa");
    CHECK(store2.branches()["main"].read_text("b.txt") == "bbb");
    CHECK(contains(store2.branches().keys(), "feature"));
    CHECK(store2.branches()["feature"].read_text("c.txt") == "ccc");

    fs::remove_all(path);
    fs::remove_all(remote_path);
    fs::remove_all(restore_path);
}

// ---------------------------------------------------------------------------
// already in sync
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: backup when already in sync", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    // Second backup should be in sync
    auto diff = store.backup(remote_url);
    CHECK(diff.in_sync());
    CHECK(diff.total() == 0);

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

// ---------------------------------------------------------------------------
// tags
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: backup with tags", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    auto remote = vost::GitStore::open(remote_path);
    auto tags = remote.tags().keys();
    CHECK(contains(tags, "v1.0"));
    CHECK(remote.tags()["v1.0"].read_text("a.txt") == "hello");

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

// ---------------------------------------------------------------------------
// bundle
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: backup to bundle", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto bundle = (path.parent_path() / "backup.bundle").string();
    auto diff = store.backup(bundle);

    CHECK_FALSE(diff.in_sync());
    CHECK_FALSE(diff.add.empty());
    CHECK(fs::exists(bundle));

    fs::remove_all(path);
    fs::remove(bundle);
}

TEST_CASE("Mirror: restore from bundle", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto bundle = (path.parent_path() / "backup.bundle").string();
    store.backup(bundle);

    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions opts;
    opts.create = true;
    auto store2 = vost::GitStore::open(restore_path, opts);

    auto diff = store2.restore(bundle);
    CHECK_FALSE(diff.in_sync());

    CHECK(contains(store2.branches().keys(), "main"));
    CHECK(store2.branches()["main"].read_text("a.txt") == "hello");
    CHECK(contains(store2.tags().keys(), "v1.0"));

    fs::remove_all(path);
    fs::remove(bundle);
    fs::remove_all(restore_path);
}

TEST_CASE("Mirror: bundle dry run", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");

    auto bundle = (path.parent_path() / "backup.bundle").string();
    vost::BackupOptions bo;
    bo.dry_run = true;
    auto diff = store.backup(bundle, bo);

    CHECK_FALSE(diff.in_sync());
    CHECK_FALSE(fs::exists(bundle));

    fs::remove_all(path);
}

TEST_CASE("Mirror: bundle round trip", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "aaa");
    f = f.write_text("b.txt", "bbb");
    store.tags().set("v1.0", f);

    auto bundle = (path.parent_path() / "roundtrip.bundle").string();
    store.backup(bundle);

    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions opts;
    opts.create = true;
    auto store2 = vost::GitStore::open(restore_path, opts);
    store2.restore(bundle);

    CHECK(store2.branches()["main"].read_text("a.txt") == "aaa");
    CHECK(store2.branches()["main"].read_text("b.txt") == "bbb");
    CHECK(contains(store2.tags().keys(), "v1.0"));

    fs::remove_all(path);
    fs::remove(bundle);
    fs::remove_all(restore_path);
}

// ---------------------------------------------------------------------------
// refs filtering
// ---------------------------------------------------------------------------

TEST_CASE("Mirror: backup with refs filter", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();

    vost::BackupOptions bo;
    bo.refs = {"main"};
    store.backup(remote_url, bo);

    auto remote = vost::GitStore::open(remote_path);
    CHECK(contains(remote.branches().keys(), "main"));
    CHECK_FALSE(contains(remote.tags().keys(), "v1.0"));

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

TEST_CASE("Mirror: restore with refs filter", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();
    store.backup(remote_url);

    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions oo;
    oo.create = true;
    auto store2 = vost::GitStore::open(restore_path, oo);

    vost::RestoreOptions ro;
    ro.refs = {"v1.0"};
    store2.restore(remote_url, ro);

    CHECK(contains(store2.tags().keys(), "v1.0"));

    fs::remove_all(path);
    fs::remove_all(remote_path);
    fs::remove_all(restore_path);
}

TEST_CASE("Mirror: backup bundle with refs", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto bundle = (path.parent_path() / "main-only.bundle").string();
    vost::BackupOptions bo;
    bo.refs = {"main"};
    store.backup(bundle, bo);

    auto restore_path = path.parent_path() / (path.filename().string() + "_restored.git");
    vost::OpenOptions oo;
    oo.create = true;
    auto store2 = vost::GitStore::open(restore_path, oo);
    store2.restore(bundle);

    CHECK(contains(store2.branches().keys(), "main"));
    CHECK_FALSE(contains(store2.tags().keys(), "v1.0"));

    fs::remove_all(path);
    fs::remove(bundle);
    fs::remove_all(restore_path);
}

TEST_CASE("Mirror: backup ref preserves existing remote refs", "[mirror]") {
    auto path = make_temp_mirror_dir();
    auto store = open_mirror_store(path);
    auto f = store.branches()["main"];
    f = f.write_text("a.txt", "hello");
    store.tags().set("v1.0", f);

    auto remote_path = path.parent_path() / (path.filename().string() + "_remote.git");
    auto remote_url = remote_path.string();

    // Full backup first
    store.backup(remote_url);

    // Verify remote has both
    {
        auto remote = vost::GitStore::open(remote_path);
        CHECK(contains(remote.branches().keys(), "main"));
        CHECK(contains(remote.tags().keys(), "v1.0"));
    }

    // Targeted backup of only main — v1.0 should still survive on remote
    vost::BackupOptions bo;
    bo.refs = {"main"};
    auto diff = store.backup(remote_url, bo);
    CHECK(diff.del.empty()); // no deletes with ref filter

    auto remote = vost::GitStore::open(remote_path);
    CHECK(contains(remote.tags().keys(), "v1.0"));

    fs::remove_all(path);
    fs::remove_all(remote_path);
}

File: rs/tests/test_mirror.rs
mod common;

use std::path::Path;
use vost::fs::WriteOptions;
use vost::types::{BackupOptions, RestoreOptions};
use vost::*;

fn create_remote_path(dir: &Path) -> String {
    dir.join("remote.git").to_string_lossy().to_string()
}

// ---------------------------------------------------------------------------
// backup
// ---------------------------------------------------------------------------

#[test]
fn backup_to_local_bare_repo() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let remote_url = create_remote_path(dir.path());
    let diff = store.backup(&remote_url, &BackupOptions::default()).unwrap();

    assert!(!diff.in_sync());
    assert!(!diff.add.is_empty());

    // Verify remote has the refs
    let remote = GitStore::open(
        &remote_url,
        OpenOptions {
            create: false,
            ..Default::default()
        },
    )
    .unwrap();
    let branches = remote.branches().list().unwrap();
    assert!(branches.contains(&"main".to_string()));
    assert_eq!(
        remote
            .branches()
            .get("main")
            .unwrap()
            .read_text("a.txt")
            .unwrap(),
        "hello"
    );
}

// ---------------------------------------------------------------------------
// restore
// ---------------------------------------------------------------------------

#[test]
fn restore_from_local_bare_repo() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Create a new empty store and restore into it
    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();

    let diff = store2.restore(&remote_url, &RestoreOptions::default()).unwrap();
    assert!(!diff.in_sync());
    assert!(!diff.add.is_empty());

    let branches = store2.branches().list().unwrap();
    assert!(branches.contains(&"main".to_string()));
    assert_eq!(
        store2
            .branches()
            .get("main")
            .unwrap()
            .read_text("a.txt")
            .unwrap(),
        "hello"
    );
}

// ---------------------------------------------------------------------------
// dry-run
// ---------------------------------------------------------------------------

#[test]
fn dry_run_backup_makes_no_changes() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let remote_url = create_remote_path(dir.path());
    // First do a real backup so remote exists
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Write more data
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("b.txt", b"world", WriteOptions::default()).unwrap();

    // Dry-run should report changes but not push
    let diff = store
        .backup(
            &remote_url,
            &BackupOptions {
                dry_run: true,
                ..Default::default()
            },
        )
        .unwrap();
    assert!(!diff.in_sync());

    // Remote should still only have the old data
    let remote = GitStore::open(
        &remote_url,
        OpenOptions {
            create: false,
            ..Default::default()
        },
    )
    .unwrap();
    assert!(!remote
        .branches()
        .get("main")
        .unwrap()
        .exists("b.txt")
        .unwrap());
}

#[test]
fn dry_run_restore_makes_no_changes() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Create empty store
    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();

    let diff = store2
        .restore(
            &remote_url,
            &RestoreOptions {
                dry_run: true,
                ..Default::default()
            },
        )
        .unwrap();
    assert!(!diff.in_sync());

    // Store2 should still be empty
    assert!(store2.branches().list().unwrap().is_empty());
}

// ---------------------------------------------------------------------------
// stale ref deletion (backup) / additive restore
// ---------------------------------------------------------------------------

#[test]
fn backup_deletes_stale_remote_refs() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    // Create a second branch
    store.branches().set("extra", &fs).unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Verify remote has both branches
    {
        let remote = GitStore::open(
            &remote_url,
            OpenOptions {
                create: false,
                ..Default::default()
            },
        )
        .unwrap();
        assert!(remote.branches().list().unwrap().contains(&"extra".to_string()));
    }

    // Delete the extra branch locally
    store.branches().delete("extra").unwrap();

    // Backup again — should delete the remote extra branch
    let diff = store.backup(&remote_url, &BackupOptions::default()).unwrap();
    assert!(diff.delete.iter().any(|r| r.ref_name.contains("extra")));

    // Verify remote no longer has the extra branch
    let remote = GitStore::open(
        &remote_url,
        OpenOptions {
            create: false,
            ..Default::default()
        },
    )
    .unwrap();
    assert!(!remote
        .branches()
        .list()
        .unwrap()
        .contains(&"extra".to_string()));
}

#[test]
fn restore_is_additive() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Create a local-only branch
    store.branches().set("local-only", &fs).unwrap();
    assert!(store
        .branches()
        .list()
        .unwrap()
        .contains(&"local-only".to_string()));

    // Restore from remote — local-only branch should survive (additive)
    let diff = store.restore(&remote_url, &RestoreOptions::default()).unwrap();
    // Diff should NOT include deletes
    assert!(diff.delete.is_empty());
    // local-only branch should still exist
    assert!(store
        .branches()
        .list()
        .unwrap()
        .contains(&"local-only".to_string()));
}

// ---------------------------------------------------------------------------
// round-trip
// ---------------------------------------------------------------------------

#[test]
fn round_trip_backup_then_restore() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"aaa", WriteOptions::default()).unwrap();
    let fs = fs.write("b.txt", b"bbb", WriteOptions::default()).unwrap();

    store.branches().set("feature", &fs).unwrap();
    let feat = store.branches().get("feature").unwrap();
    let _feat = feat
        .write("c.txt", b"ccc", WriteOptions::default())
        .unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Create new store and restore
    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();
    store2.restore(&remote_url, &RestoreOptions::default()).unwrap();

    assert_eq!(
        store2
            .branches()
            .get("main")
            .unwrap()
            .read_text("a.txt")
            .unwrap(),
        "aaa"
    );
    assert_eq!(
        store2
            .branches()
            .get("main")
            .unwrap()
            .read_text("b.txt")
            .unwrap(),
        "bbb"
    );
    let branches = store2.branches().list().unwrap();
    assert!(branches.contains(&"feature".to_string()));
    assert_eq!(
        store2
            .branches()
            .get("feature")
            .unwrap()
            .read_text("c.txt")
            .unwrap(),
        "ccc"
    );
}

// ---------------------------------------------------------------------------
// already in sync
// ---------------------------------------------------------------------------

#[test]
fn backup_when_already_in_sync() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    // Second backup should be in sync
    let diff = store.backup(&remote_url, &BackupOptions::default()).unwrap();
    assert!(diff.in_sync());
    assert_eq!(diff.total(), 0);
}

// ---------------------------------------------------------------------------
// tags
// ---------------------------------------------------------------------------

#[test]
fn backup_with_tags() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();
    store.tags().set("v1.0", &fs).unwrap();

    let remote_url = create_remote_path(dir.path());
    store.backup(&remote_url, &BackupOptions::default()).unwrap();

    let remote = GitStore::open(
        &remote_url,
        OpenOptions {
            create: false,
            ..Default::default()
        },
    )
    .unwrap();
    let tags = remote.tags().list().unwrap();
    assert!(tags.contains(&"v1.0".to_string()));
    assert_eq!(
        remote
            .tags()
            .get("v1.0")
            .unwrap()
            .read_text("a.txt")
            .unwrap(),
        "hello"
    );
}

// ---------------------------------------------------------------------------
// bundle
// ---------------------------------------------------------------------------

#[test]
fn backup_to_bundle() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();
    store
        .tags()
        .set("v1.0", &store.branches().get("main").unwrap())
        .unwrap();

    let bundle_path = dir
        .path()
        .join("backup.bundle")
        .to_string_lossy()
        .to_string();
    let diff = store
        .backup(&bundle_path, &BackupOptions::default())
        .unwrap();

    assert!(!diff.in_sync());
    assert!(!diff.add.is_empty());
    assert!(std::path::Path::new(&bundle_path).exists());
}

#[test]
fn restore_from_bundle() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();
    store
        .tags()
        .set("v1.0", &store.branches().get("main").unwrap())
        .unwrap();

    let bundle_path = dir
        .path()
        .join("backup.bundle")
        .to_string_lossy()
        .to_string();
    store
        .backup(&bundle_path, &BackupOptions::default())
        .unwrap();

    // Create a new empty store and restore from bundle
    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();

    let diff = store2
        .restore(&bundle_path, &RestoreOptions::default())
        .unwrap();
    assert!(!diff.in_sync());

    let branches = store2.branches().list().unwrap();
    assert!(branches.contains(&"main".to_string()));
    assert_eq!(
        store2
            .branches()
            .get("main")
            .unwrap()
            .read_text("a.txt")
            .unwrap(),
        "hello"
    );
    let tags = store2.tags().list().unwrap();
    assert!(tags.contains(&"v1.0".to_string()));
}

#[test]
fn bundle_dry_run() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let _fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();

    let bundle_path = dir
        .path()
        .join("backup.bundle")
        .to_string_lossy()
        .to_string();
    let diff = store
        .backup(
            &bundle_path,
            &BackupOptions {
                dry_run: true,
                ..Default::default()
            },
        )
        .unwrap();

    assert!(!diff.in_sync());
    assert!(!std::path::Path::new(&bundle_path).exists());
}

#[test]
fn bundle_round_trip() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"aaa", WriteOptions::default()).unwrap();
    let _fs = fs.write("b.txt", b"bbb", WriteOptions::default()).unwrap();
    store
        .tags()
        .set("v1.0", &store.branches().get("main").unwrap())
        .unwrap();

    let bundle_path = dir
        .path()
        .join("roundtrip.bundle")
        .to_string_lossy()
        .to_string();
    store
        .backup(&bundle_path, &BackupOptions::default())
        .unwrap();

    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();
    store2
        .restore(&bundle_path, &RestoreOptions::default())
        .unwrap();

    assert_eq!(
        store2
            .branches()
            .get("main")
            .unwrap()
            .read_text("a.txt")
            .unwrap(),
        "aaa"
    );
    assert_eq!(
        store2
            .branches()
            .get("main")
            .unwrap()
            .read_text("b.txt")
            .unwrap(),
        "bbb"
    );
    assert!(store2
        .tags()
        .list()
        .unwrap()
        .contains(&"v1.0".to_string()));
}

// ---------------------------------------------------------------------------
// refs filtering
// ---------------------------------------------------------------------------

#[test]
fn backup_with_refs_filter() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();
    store.tags().set("v1.0", &fs).unwrap();

    let remote_url = create_remote_path(dir.path());
    let opts = BackupOptions {
        refs: Some(vec!["main".to_string()]),
        ..Default::default()
    };
    store.backup(&remote_url, &opts).unwrap();

    let remote = GitStore::open(
        &remote_url,
        OpenOptions {
            create: false,
            ..Default::default()
        },
    )
    .unwrap();
    let branches = remote.branches().list().unwrap();
    assert!(branches.contains(&"main".to_string()));
    // Tag should NOT have been pushed
    let tags = remote.tags().list().unwrap();
    assert!(!tags.contains(&"v1.0".to_string()));
}

#[test]
fn restore_with_refs_filter() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();
    store.tags().set("v1.0", &fs).unwrap();

    let remote_url = create_remote_path(dir.path());
    store
        .backup(&remote_url, &BackupOptions::default())
        .unwrap();

    // Create new store
    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();

    // Only restore the tag
    let opts = RestoreOptions {
        refs: Some(vec!["v1.0".to_string()]),
        ..Default::default()
    };
    store2.restore(&remote_url, &opts).unwrap();

    let tags = store2.tags().list().unwrap();
    assert!(tags.contains(&"v1.0".to_string()));
}

#[test]
fn backup_bundle_with_refs() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs.write("a.txt", b"hello", WriteOptions::default()).unwrap();
    store.tags().set("v1.0", &fs).unwrap();

    let bundle_path = dir
        .path()
        .join("main-only.bundle")
        .to_string_lossy()
        .to_string();
    let opts = BackupOptions {
        refs: Some(vec!["main".to_string()]),
        ..Default::default()
    };
    store.backup(&bundle_path, &opts).unwrap();

    // Restore bundle into new store — should only have main, not the tag
    let store2 = GitStore::open(
        dir.path().join("restored.git"),
        OpenOptions {
            create: true,
            branch: None,
            ..Default::default()
        },
    )
    .unwrap();
    store2
        .restore(&bundle_path, &RestoreOptions::default())
        .unwrap();

    let branches = store2.branches().list().unwrap();
    assert!(branches.contains(&"main".to_string()));
    let tags = store2.tags().list().unwrap();
    assert!(!tags.contains(&"v1.0".to_string()));
}
